protofsm/state_machine.go

```diff
@@ -0,0 +1,428 @@
+
+			for {
+				select {
+				case <-predicateTicker.C:
+					canSend := fn.MapOptionZero(
+						daemonEvent.SendWhen,
+						func(pred SendPredicate) bool {
+							return pred()
+						},
+					)
+
+					if canSend {
+						sendAndCleanUp()
+						return
+					}
+
+				case <-s.quit:
+					return
+				}
+			}
+		}()
+
+		return nil
+
+	// If this is a disable channel event, then we'll disable the channel.
+	// This is usually done for things like co-op closes.
+	case *DisableChannelEvent:
+		err := s.daemon.DisableChannel(daemonEvent.ChanPoint)
+		if err != nil {
+			return fmt.Errorf("unable to disable channel: %w", err)
+		}
+
+		return nil
+
+	// If this is a broadcast transaction event, then we'll broadcast with
+	// the label attached.
+	case *BroadcastTxn:
+		err := s.daemon.BroadcastTransaction(
+			daemonEvent.Tx, daemonEvent.Label,
+		)
+		if err != nil {
+			// TODO(roasbeef): hook has channel read event event is
+			// hit?
+			return fmt.Errorf("unable to broadcast txn: %w", err)
+		}
+
+		return nil
+	}
+
+	return fmt.Errorf("unknown daemon event: %T", event)
+}
+
+// applyEvents applies a new event to the state machine. This will continue
+// until no further events are emitted by the state machine. Along the way,
+// we'll also ensure to execute any daemon events that are emitted.
+func (s *StateMachine[Event, Env]) applyEvents(newEvent Event) (State[Event, Env], error) {
+	// TODO(roasbeef): make starting state as part of env?
+	currentState := s.currentState
+
+	eventQueue := fn.NewQueue(newEvent)
+
+	// Given the next event to handle, we'll process the event, then add
+	// any new emitted internal events to our event queue. This continues
+	// until we reach a terminal state, or we run out of internal events to
+	// process.
+	for nextEvent := eventQueue.Dequeue(); nextEvent.IsSome(); nextEvent = eventQueue.Dequeue() {
+		err := fn.MapOptionZero(nextEvent, func(event Event) error {
+			// Apply the state transition function of the current
+			// state given this new event and our existing env.
+			transition, err := currentState.ProcessEvent(
+				event, s.env,
+			)
+			if err != nil {
+				return err
+			}
+
+			newEvents := transition.NewEvents
+			err = fn.MapOptionZero(newEvents, func(events EmittedEvent[Event]) error {
+				// With the event processed, we'll process any
+				// new daemon events that were emitted as part
+				// of this new state transition.
+				err := fn.MapOptionZero(events.ExternalEvents, func(dEvents DaemonEventSet) error {
+					for _, dEvent := range dEvents {
+						err := s.executeDaemonEvent(dEvent)
+						if err != nil {
+							return err
+						}
+					}
+
+					return nil
+				})
+				if err != nil {
+					return err
+				}
+
+				// Next, we'll add any new emitted events to
+				// our event queue.
+				events.InternalEvent.WhenSome(func(inEvent Event) {
+					eventQueue.Enqueue(inEvent)
+				})
+
+				return nil
+			})
+			if err != nil {
+				return err
+			}
```

Comment:
If I'm reading this correctly it means that external events will always synchronize before internal ones. I believe this is OK (even preferred), but wanted to double check that this is what we want. Is there ever a situation in which we'd want to emit both but have them synchronize the other direction?

I think Internal events are a way to buy us "cut-through" where we don't rely on synthetic events from the surrounding environment to drive ourselves forward and we opportunistically drive ourselves as far forward as possible in any given moment. Under this interpretation I think this is indeed what we want.

Reply:
>  Is there ever a situation in which we'd want to emit both but have them synchronize the other direction?

Good q....my mental model here is that the executor context switches to execute all the syscalls, then resumes execution of the state machine with any emitted internal events. Re the opposite ordering, I guess this is sort of a system call interface the executor has with the FSM: what's the execution order of emitted events? One could likely devise state machines where if you flip the ordering you may end up with incorrect (?) behavior.

Reply:
> I think Internal events are a way to buy us "cut-through" where we don't rely on synthetic events from the surrounding environment to drive ourselves forward and we opportunistically drive ourselves as far forward as possible in any given moment. 

Yep, I think for the most part, you can elide emitting an internal event just by doing even _more_ within a given state transition. I like them though from the PoV of minimal state transitions, and also domain modeling as well.